codefix: Fix code actions that return Command[] directly instead of CodeAction[]

liskin opened the pull request:
According to
https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#textDocument_codeAction,
the response to textDocument/codeAction is:

    (Command | CodeAction)[] | null

and the code only handled the case where it was a CodeAction that either
specified an edit or a command, but didn't handle a direct Command.

Note that the specification also says that both can be specified and
then the edit is applied first, then the command. Furthermore, there
seems to be some hacky code handling arguments directly, which I suspect
is non-standard and only works with a specific LSP server that happens
to pass the edits in the arguments unmodified.

<!--
Before creating a pull request, do the following.

* Read the Contributing guide linked above first.
* Read the documentation that comes with ALE with `:help ale-dev`.

Have fun!
-->

> Where are the tests? Have you added tests? Have you updated the tests? Read the
comment above and the documentation referenced in it first. Write tests!

There were no other tests for this part of ALE, and I haven't had time to add them yet. I'd like to ping @daliusd, who's the original author of this code, to check it and perhaps shed some light on the last paragraph of the commit message, i.e. why the code looks into arguments even though the LSP protocol spec doesn't look like editors are supposed to touch arguments directly. I am not opposed to writing tests, I'm a bit time-constrained these days however, so I'd rather put at least something out there now.

liskin commented:
Oh, there actually are tests. Mea culpa. Will look at this later then.

daliusd commented:
I wrote LSP part for fun, I don't use any LSP myself (only tsserver which is not really LSP) so there might be missing parts. Do you have any sample response?

daliusd commented on autoload/ale/codefix.vim:
```diff
@@ -201,6 +201,13 @@ function! ale#codefix#ApplyLSPCodeAction(data, item) abort
         \)
 
+        let l:request_id = ale#lsp#Send(a:data.connection_id, l:message)
+    elseif type(a:item.command) == v:t_string
```
you must check `has_key(a:item, 'command')` as well for readability at least. I'm not sure what vimscript does if there is no item in dict `a:item`.

liskin commented:
> I wrote LSP part for fun, I don't use any LSP myself (only tsserver which is not really LSP) so there might be missing parts. Do you have any sample response?

Oh, that was my situation as well with my first code actions PR a year ago, but now I started using https://github.com/python-lsp/python-lsp-server and https://github.com/haskell/haskell-language-server somewhat regularly and wanted to make them work. :-)

I'm attaching a sample response from haskell-language-server. I think I've fixed the test and will push a fixup in a minute as well.

```json
{
  "result": [
    {
      "title": "Disable \"typed-holes\" warnings",
      "command": "859695:hls:fallbackCodeAction",
      "arguments": [
        {
          "fallbackCommand": null,
          "fallbackWorkspaceEdit": {
            "changes": {
              "file:///home/tomi/.xmonad-testing/xmonad-contrib/XMonad/Hooks/StatusBar/PP.hs": [
                {
                  "range": {
                    "end": {
                      "character": 0,
                      "line": 3
                    },
                    "start": {
                      "character": 0,
                      "line": 3
                    }
                  },
                  "newText": "{-# OPTIONS_GHC -Wno-typed-holes #-}\n"
                }
              ]
            }
          }
        }
      ]
    },
    {
      "title": "Wingman: Attempt to fill hole",
      "command": "859695:hls:fallbackCodeAction",
      "arguments": [
        {
          "fallbackCommand": {
            "title": "Wingman: Attempt to fill hole",
            "command": "859695:tactics:tacticsAutoCommand",
            "arguments": [
              {
                "tp_range": {
                  "end": {
                    "character": 7,
                    "line": 265
                  },
                  "start": {
                    "character": 6,
                    "line": 265
                  }
                },
                "tp_var_name": "",
                "tp_file": "file:///home/tomi/.xmonad-testing/xmonad-contrib/XMonad/Hooks/StatusBar/PP.hs"
              }
            ]
          },
          "fallbackWorkspaceEdit": null
        }
      ]
    },
    {
      "title": "Wingman: Case split on a",
      "command": "859695:hls:fallbackCodeAction",
      "arguments": [
        {
          "fallbackCommand": {
            "title": "Wingman: Case split on a",
            "command": "859695:tactics:tacticsDestructCommand",
            "arguments": [
              {
                "tp_range": {
                  "end": {
                    "character": 7,
                    "line": 265
                  },
                  "start": {
                    "character": 6,
                    "line": 265
                  }
                },
                "tp_var_name": "a",
                "tp_file": "file:///home/tomi/.xmonad-testing/xmonad-contrib/XMonad/Hooks/StatusBar/PP.hs"
              }
            ]
          },
          "fallbackWorkspaceEdit": null
        }
      ]
    },
    {
      "title": "Wingman: Split all function arguments",
      "command": "859695:hls:fallbackCodeAction",
      "arguments": [
        {
          "fallbackCommand": {
            "title": "Wingman: Split all function arguments",
            "command": "859695:tactics:tacticsDestructAllCommand",
            "arguments": [
              {
                "tp_range": {
                  "end": {
                    "character": 7,
                    "line": 265
                  },
                  "start": {
                    "character": 6,
                    "line": 265
                  }
                },
                "tp_var_name": "",
                "tp_file": "file:///home/tomi/.xmonad-testing/xmonad-contrib/XMonad/Hooks/StatusBar/PP.hs"
              }
            ]
          },
          "fallbackWorkspaceEdit": null
        }
      ]
    },
    {
      "title": "Wingman: Refine hole",
      "command": "859695:hls:fallbackCodeAction",
      "arguments": [
        {
          "fallbackCommand": {
            "title": "Wingman: Refine hole",
            "command": "859695:tactics:tacticsRefineCommand",
            "arguments": [
              {
                "tp_range": {
                  "end": {
                    "character": 7,
                    "line": 265
                  },
                  "start": {
                    "character": 6,
                    "line": 265
                  }
                },
                "tp_var_name": "",
                "tp_file": "file:///home/tomi/.xmonad-testing/xmonad-contrib/XMonad/Hooks/StatusBar/PP.hs"
              }
            ]
          },
          "fallbackWorkspaceEdit": null
        }
      ]
    },
    {
      "title": "Wingman: Use custom tactic block",
      "command": "859695:hls:fallbackCodeAction",
      "arguments": [
        {
          "fallbackCommand": {
            "title": "Wingman: Use custom tactic block",
            "command": "859695:tactics:tacticsBeginMetaprogramCommand",
            "arguments": [
              {
                "tp_range": {
                  "end": {
                    "character": 7,
                    "line": 265
                  },
                  "start": {
                    "character": 6,
                    "line": 265
                  }
                },
                "tp_var_name": "",
                "tp_file": "file:///home/tomi/.xmonad-testing/xmonad-contrib/XMonad/Hooks/StatusBar/PP.hs"
              }
            ]
          },
          "fallbackWorkspaceEdit": null
        }
      ]
    }
  ],
  "id": 6,
  "jsonrpc": "2.0"
}
```

liskin commented on test/test_codefix.vader:
```diff
@@ -444,16 +444,6 @@ Execute("workspace/applyEdit" from LSP should be handled):
-  AssertEqual g:handle_code_action_called, 1
-  AssertEqual
-  \ [{'description': 'codeaction', 'changes': [{'fileName': '/foo/bar/file.ts', 'textChanges': [{'end': {'offset': 32, 'line': 3}, 'newText': ', createVideo', 'start': {'offset': 32, 'line': 3}}]}]}],
-  \ g:code_actions
```
I don't think this test made any sense:

1. there's just one action even though the test description says otherwise
2. it's a tsserver response so why is this being handled by `HandleLSPResponse`

Anyway I used most of it for the new test, just changed the expectation to what I think should happen, and what works with more standard language servers.

daliusd replied:
OK

liskin commented:
Cool, thanks. I squashed the fixup and forcepushed.
